---
title: "Processing FASTQ Files"
author: "Rui"
date: "2025-02-20"
output: html_document
---

# Introduction

This document is meant to serve as a guide for processing bulk rna-seq data. The starting point of this guide is fastq files and the end result 
will be a gene count matrix. This guide uses bash and R languages, and although not intended to be a tutorial in either, will take extra time to 
explain and show how code works and why. Additionally, this guide makes use of several open source (i.e., free) softwares that will need to be 
downloaded. The software described in this guide are not the only ones available, there are plently that could be substituted for each step.

This guide assumes that each fastq file only contains data from one biological sample. The data presented here will be paired end. If you have multiple samples
 in the same fastq file (this is called multiplexed), or both read pairs in the same fastq file (interleaved), then in order to follow this guide you must 
 de-multiplex and de-interleave your 

Fastq files can be very large, several GB each or larger. You will have to consider your computer's storage capabilities. All of the steps in this analysis will 
require some file(s) as input and produce some other file(s) as output. The output files can also be large. For example, quality trimming of fastq files will 
remove low quality reads/bases, and the output of this process is new fastq files, which, if your data is of high quality to begin with, will be only slightly 
smaller than the original files. 

Additionally, this process can be very resource intensive, requiring a lot of RAM. Your computer should have ample storage space and at least 64GB of RAM. 
Not all computers are suitable for this process.

## Overview

The general procedure is as follows:

Fastqs should be downloaded onto your computer, there are surely ways to perform this type of analysis without having the files stored locally but that is 
beyond the scope of this guide. Usually, whoever created the fastq files (sequencing company, etc) will provide you with either an FTP (file transfer protocol) 
link or a physical hard drive. If you have a physical hard drive containing the files, simply COPY them onto your computer, and maintain the original files on 
the hard drive as a backup. If you have to download them through an FTP, follow the FTP instructions. Always be sure to maintain a backup of the fastq files, 
incase you ever need them. 

Fastq files should be accompanied by an MD5 file. Usually there is one MD5 file for each fastq file. An MD5 file is a checksum file that verifies the file's 
data is intact. During transfers of large amount of data it is possible that something goes wrong, and the downloaded data is corrupted, even if only slighly. 
You should create your own MD5 checksums for each fastq file and verify that it matches the one provided exactly. If not, try redownloading the corrupted files 
and checksum again. Do this immediately, especially if downloaded through an FTP, as some providers will only keep the files for a short time. 

Next, you should evaluate the quality of the fastq files using fastqc. This software will produce one html file for each fastq file (that you will open with your
 web browser) showing the average data quality for all the reads in the file, any adapter content, and several other things. You can use this information to 
inform your trimming step, which is next.

Now you will trim your fastq files using trimmomatic. You can tell trimmomatic to trim low quality reads and bases according to your needs, and also use it to 
trim adapter sequences. This step will result in new, trimmed fastq files. 

The next step is mapping the trimmed fastq files to a reference genome using STAR. STAR requires its own special reference which must be created by supplying 
a fasta (.fa) and gene annotation (.gtf) file. After creating the reference, we use it to map the reads. The result of this step is bam files (could be sam files 
but I recommend bam). 

Then, we use featureCounts to create gene counts from the bam files. This step requires that we supply a gene annotation file (.gtf, the same one used when creating 
the STAR reference) and a bam file. The result is a file that contains the counts for each gene from that bam file, which can be in .txt or .csv or whatever 
format. 

Finally, the output from featureCounts can be processed using something like R to make a gene matrix. 

# Part 0: Getting set up and a few things about bash

In order to keep your analysis organized, you should make a new folder on your computer dedicated to the experiment you are working with. Additionally, 
you should have subfolders to keep different steps of the process separate. For example, at the beginning I have a main folder called "rnaseq_tutorial", and
within it I have a folder called "fastq" that contains all the fastq files for this tutorial. I'll make new folders for each of the resulting files as we go, but
code files and scripts can stay just in "rnaseq_tutorial". 

For the purposes of this tutorial, I've created dummy fastq files by taking the first 100 reads of a cohort of files. This is mostly just to demonstrate
how to work with fastq files, we won't actually process any of them. 

Now a few basics about bash. I'll show some of the common things we'll need to do.

To make a new folder use the mkdir (make directory) command

```{bash eval = FALSE}
mkdir new_folder
```

To make a new folder and subfolders at the same time, use mkdir -p 

```{bash eval = FALSE}
mkdir -p new_folder/new_subfolder
```

We could achieve the same result by doing this: 

```{bash eval = FALSE}
mkdir new_folder
mkdir new_folder/new_subfolder
```

But using mkdir -p is quicker.

The echo command is also useful. echo in bash is akin to print in other programming languages.

Like this:

```{bash}
echo "Hello World"
```

We can also do this:

```{bash}
echo fastq/*
```

This is all the fastq files in the fastq directory. The * in fastq/* means "everything". i.e., echo (print) everything inside the fastq folder.

We can also use the * with other characters to be more specific.

```{bash}
echo fastq/*_1.fq 
```

This printed all the _1.fq files (all the forward reads). You will see more examples of this as we go.

Next we'll look at for loops, which will be used a lot. The basic idea of a for loop is as follows: 

```{bash eval = FALSE}
for Each_Item in Some_List; do

    Some_Task

done
```

For example: 

```{bash}
for file in fastq/*; do
    echo $file
done
```

The "\$" in "$file" tells bash that "file" is a variable.

Now if we didnt want the entire file path when printing all the file names, we can use the basename command:

```{bash}
for file in fastq/*; do
    echo $(basename $file)
done
```

We can also use basename to remove parts of the file name:

```{bash}
for file in fastq/*; do
    echo $(basename $file .fastq.gz)
done
```

There is much more we could demonstrate, but for now that is all. We will proceed with the analysis.

# Part1: fastqc for quality control insights

Fastqc (https://www.bioinformatics.babraham.ac.uk/projects/download.html#fastqc) is a program that produces reports about fastq data. From these reports we can
determine the overall quality of our data and how aggressively we need to trim it. 

Fastqc is easy to use, we don't even have to use a for loop. We can simply pass a list (fastq/* is a list) and it will handle the rest.

Before we use it, we'll make a place to put it's output:

```{bash}
mkdir -p quality_report/before_trimming
```

Notice that we're putting these reports in a before_trimming subfolder... We will run fastqc again after trimming to make sure we trimmed well enough. 

Now to run fastqc. We use the --outdir argument to tell fastqc where to put the output. We'll also use the --quiet argument to suppress all progress messages in order to
 save some page space. I'll add some of what the output would be so you can see. 

```{bash eval = FALSE}
fastqc fastq/* --outdir=quality_report/before_trimming --quiet
```

application/gzip
application/gzip
Started analysis of Sample-312_2024-028-1-S4_R1.fastq.gz
application/gzip
application/gzip
application/gzip
application/gzip
application/gzip
application/gzip
Approx 5% complete for Sample-312_2024-028-1-S4_R1.fastq.gz
Approx 10% complete for Sample-312_2024-028-1-S4_R1.fastq.gz
... etc. 

Also, you could run fastq -h to see the help menu and all arguments/options.

You can see each of the fastqc reports here https://ruitreves.github.io/rnaseq_tutorial/

These fastq files were specifically chosen for this tutorial because they are a bit worse than others from the same cohort. 

Now we will move on to the trimming step. 

# Part2: Trimmomatic 

We will use trimmomatic http://www.usadellab.org/cms/?page=trimmomatic to remove adapter sequences and trim low quality bases from our reads. With trimmomatic,
 we do need to use a for loop, and you will see why in a moment. Also, this step will demonstrate how using a software that is not in your path is different than
 using one that is in your path. 

We will be using a lot more arguments for trimmomatic than we did for fastqc, and sometimes the code we have to write gets pretty long. You can read all about 
trimmomatic in its manual, and we'll only briefly describe the ones we use here. 

This is what we want to do: 

```{bash eval = FALSE}
java -jar /Users/hardestylab4/Desktop/star/Trimmomatic-0.39/trimmomatic-0.39.jar \
	PE -threads 16 -phred33 \
	fastq/Sample-312_2024-028-1-S4_R1.fastq.gz fastq/Sample-312_2024-028-1-S4_R2.fastq.gz \
	trimmed_fastq/trimmed_pairedSample-312_2024-028-1-S4_R1.fastq.gz trimmed_fastq/trimmed_unpairedSample-312_2024-028-1-S4_R1.fastq.gz \
	trimmed_fastq/trimmed_pairedSample-312_2024-028-1-S4_R2.fastq.gz trimmed_fastq/trimmed_unpairedSample-312_2024-028-1-S4_R2.fastq.gz \
	ILLUMINACLIP:/Users/hardestylab4/Desktop/star/Trimmomatic-0.39/adapters/TruSeq3-PE-2.fa:2:30:10:2:TRUE LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 \
	MINLEN:36
```

Theres a lot going on there. To fully understand it you will need to read the manual, or at least parts of it. We'll only discuss some parts here. 

Note: by putting a "\\" at the end of a line, we tell the terminal to expect more input for the same command but on a new line. This can make writing long 
commands easier, since we dont have to write them all on the same line. 

In line 1: trimmomatic is a java application, so we fist have to call java in order to run it. Thats why we start with "java".
A jar file is like a java executable, i.e., the thing were trying to run, trimmomatic-0.39.jar. By writing -jar were telling java to expect a .jar file.
Then, we simply write the path to the trimmomatic-0.39.jar file, /Users/hardestylab4/Desktop/star/Trimmomatic-0.39/trimmomatic-0.39.jar. 

That last part is the difference with having something in your path or not. trimmomatic-0.39.jar is not in my path, so I have to tell the computer where it is. 
Simply writing java -jar trimmomatic-0.39.jar wont work unless trimmomatic-0.39.jar is in the same directory I'm writing in. 
On the other hand, fastqc is in my path, so I dont have to write the path/to/fastqc, I only have to write fastqc. Then my computer looks for it in the current 
directory, doesnt find it, checks my path, and finds it there. 

In line 2: I tell trimmomatic to run in paired-end mode since our data is paired-end with "PE". then i specify -threads 16 to tell trimmomatic to use 16 threads,
which is like computing power. Some computers have less threads. -phred33 tells trimmomatic that the quality of each base in the fastq files is encoding in phred33,
which is a type of quality encoding used for fastqc files. some fastq files use phred64. 

in line 3: This is the input for trimmomatic. I first give the forward read (R1) and then the reverse read (R2).

in line 4: this is the first part of the output. trimmed_fastq/trimmed_pairedSample-312_2024-028-1-S4_R1.fastq.gz tells trimmomatic where to put the reads with paired 
output from the R1 file, and trimmed_fastq/trimmed_unpairedSample-312_2024-028-1-S4_R1.fastq.gz the unpaired output from R1. We don't really care about the unpaired
output, and we wont use it later. For the following steps, we need to have all of our reads belong to a pair. 

in line 5: this is the same as line 4, except for R2. 

in line 6: we use trimmomatic's ILLUMINACLIP argument to trim adapter sequences. We specify the path to a premade TruSeq-PE-2.fa adapter file. This file comes with
trimmomatic and contains known Illumina adapter sequences. You can use other adapter files and even make your own custom one. The rest tells trimmomatic how to 
look for adapter sequences and what to consider low quality for trimming. 

in line 7: We specify the minimum length of sequences to be kept, 36. 

I will say it again, refer to the manual for a deeper understanding of everything happening here. 

Now, lets talk about why we're going to use a for loop for this part. 

The above code is for one sample, one pair of fastq files. We have 8 fastq files (4 pairs) that were processing here, but imagine if we had 20, or more. 
We would have to write all the above code lots of times, which is annoying and inefficient. Sure, we could copy and paste and put the correct file names in each time,
but this is also inefficient and opens us up to the possibility of making errors. By using a for loop, we can be sure that the code will be the same every time 
and the files processed will be the correct ones. So lets build the code we want to use, with the bit above as a guide. 

Using our knowledge of for loops from part 0, we know we can do this:

```{bash}
for file in fastq/*; do
    echo $file
done
```

But, we want to include each pair in the same command since they get processed together. So instead of looping through each file, we actually want to loop through
each pair of files. Like this:

```{bash}
for pair in fastq/*R1.fastq.gz; do
    echo $pair
done
```

There, now we have all the R1s. But how do we get the R2s, too? 

```{bash}
for pair in fastq/*R1.fastq.gz; do
    pair_name=${pair//_R1.fastq.gz}
    echo $pair_name
done
```

This is called parameter expansion, where we basically trim off parts of a string we dont want. Now we can add the R1 and R2 parts back ourselves, thus creating 
the inputs (and output) that we need. 

Note that we removed the underscore (_) at the end of the pair_name variable, so we need to write it again. This is technically inefficient but it increases the
readability, in my opinion. 



```{bash}
for pair in fastq/*R1.fastq.gz; do
    pair_name=${pair//_R1.fastq.gz}
    echo $pair_name

    r1=${pair_name}_R1.fastq.gz
    r2=${pair_name}_R2.fastq.gz

    echo $r1
    echo $r2
done
```

That is what we need. 