---
title: "Processing FASTQ Files"
author: "Rui"
date: "2025-02-20"
output: html_document
---

# Introduction

This document is meant to be a guide for analyzing bulk RNA-seq data. The starting point of this guide will be fastq files and the end will be gene 
level counts. Various softwares will be used for this analysis, which are downloaded on, and in some cases available in the $path of this computer.

The way this guide will work is as follows:

First we will see a script of the entire process (or pipeline, if you like). 

Next we will look individually at the steps of the process.

Last we will dig into some of the code used during the process.

Feel free to jump to whichever section you need.

# Part 1: The Entire Process

```{bash eval = FALSE}
#!/bin/bash

# specify where fastq files are located. dont include a "/" at the end

fastq_location="fastq"

mkdir -p quality_report/before_trimming

# run fastqc

fastqc ${fastq_location}/* --outdir quality_report/before_trimming

# then use a for loop to trim all fastq file with trimmomatic 

mkdir trimmed_fastq

for file in ${fastq_location}/*R1.fastq.gz; do
    filename=$(basename $file "_R1.fastq.gz")
    new_filename=${filename//_2024*}
    java -jar /Users/hardestylab4/Desktop/star/Trimmomatic-0.39/trimmomatic-0.39.jar \
	PE -threads 16 -phred33 \
	fastq/${filename}_R1.fastq.gz fastq/${filename}_R2.fastq.gz \
	trimmed_fastq/trimmed_paired${new_filename}_R1.fastq.gz trimmed_fastq/trimmed_unpaired${new_filename}_R1.fastq.gz \
	trimmed_fastq/trimmed_paired${new_filename}_R2.fastq.gz trimmed_fastq/trimmed_unpaired${new_filename}_R2.fastq.gz \
	ILLUMINACLIP:/Users/hardestylab4/Desktop/star/Trimmomatic-0.39/adapters/TruSeq3-PE-2.fa:2:30:10:2:TRUE LEADING:3 TRAILING:3 SLIDINGWINDOW:4:15 \
	MINLEN:36
done

# run fastqc again to see the effects of trimming

mkdir quality_report/after_trimming

fastqc trimmed_fastq/trimmed_paired* --outdir=quality_report/after_trimming

# now run star 

mkdir bams

for file in trimmed_fastq/trimmed_paired*R1.fastq.gz; do
    filename=$(basename $file "_R1.fastq.gz")
    sample=${filename//trimmed_paired}
    echo $sample
    STAR --runThreadN 16 \
        --readFilesIn trimmed_fastq/${filename}_R1.fastq.gz trimmed_fastq/${filename}_R2.fastq.gz \
        --readFilesCommand gunzip -c \
        --genomeDir /Users/hardestylab4/Desktop/star/data/index_39 \
        --outSAMtype BAM SortedByCoordinate \
        --outFileNamePrefix bams/$sample
done

# then run featureCounts

mkdir counts

for file in bams/*.bam; do 
    filename=$(basename $file)
    sample=${filename//Aligned*}
    featureCounts --largestOverlap -a /Users/hardestylab4/Desktop/star/data/ref/Mus_musculus.GRCm39.112.gtf \
    -t exon -g gene_id -s 0 -p --countReadPairs -T 8 \
    -o counts/${sample}_counts.txt $file
done
```

That is the entire analysis from fastq files -> gene counts. There is further data processing needed to turn the current gene counts into a 
coherent matrix, but this gets us most of the way done. If all you need is to plug your fastq directory in and tweak a few things, then you're 
done with this guide. If you would like a further breakdown of each step of the process, read on. 

# Part 2: Breakdown of the steps

In this section we will breakdown each step of the process in some level of detail. 

## Step 0: Setting Up Your Workspace

Keeping things organized will make your analysis much smoother. Whether you're doing this for the first time or the hundreth time, at some point 
disorganization will cause you problems. The method used in this guide involves having separate directories (folders) dedicated to each step of 
the process. 

For example, this project (i.e., analyzing this cohort (study, experiment, etc.)) was started in its own brand new directory. A subdirectory was 
created for the raw fastq files and a code file was created to write the analysis code. There wasnt anything other than these two things in the project
folder in order to minimize clutter. As the pipeline arrived at each step of the process, a new directory was created within the main project 
directory dedicated to the output of that step. This approach keeps the workspace clean and organized, but is not a requirement, only a recommendation.

Now, a word about the softwares used here. 

Each of the softwares used in this analysis are open source (free) and created by other researchers. These softwares are different than others that 
most people use in that they are used only through the command line. That is, they don't have a graphical user interface (actually, fastqc does) where
you open up the software and click around in order to use it. Instead, you use the command line to first "call" the software and then tell it what 
you want it to do. By calling the software, I mean starting it up, evoking it, telling it to run, etc. 

Take fastqc as an example. The way its used in the script in Part 1 is:

```{bash eval = FALSE}
fastqc ${fastq_location}/* --outdir quality_report/before_trimming
```

The first part: fastqc, is us telling fastqc to run. The parts that come after are us telling fastqc what to do. 

Another thing to note is that if you go to your computer's command line and type "fastqc", probably nothing will happen, unless of course you already
have fastqc downloaded and accessible via your path, or you're command prompt is already at the folder containing the fastqc executable. 



All that to say, you need to download each of these softwares and if you want to be able to call them from "anywhere" on your computer (like in a 
specific project directory where your fastq files are) you will need to put them somewhere that your computer knows where to look. Your computer will
look for the software you're trying to call in the current directory you're command prompt is in, or it will look in your path ($PATH). You can also 
use the absolute filepath to access softwares not in your path or in your current directory. I did this in the script for trimmomatic:

```{bash eval = FALSE}
java -jar /Users/hardestylab4/Desktop/star/Trimmomatic-0.39/trimmomatic-0.39.jar
```

Trimmomatic is not in my path, so I had to use the entire filepath to call it. If I didnt, I would get an error. See below.

```{bash error = TRUE}
trimmomatic-0.39.jar
```

We get an error here because my computer doesn't know where to find trimmomatic-0.39.jar unless I specify where it is, since its not in my path. 

You can see what your path is by doing the following in bash:

```{bash}
echo $PATH
```

As you can see, there are actually many places in my path (they are separated by a ":"). You should read about how to configure and add things to your 
path and BE CAREFUL! You can disrupt important processes on your computer if you unintentionally delete, move, or change things in your path. You 
should be sure of what you're doing before you remove anything from your path or do anything to change it. 

## Step 1: Running fastqc

